### DISCUSSION TITLE
How to handle subvolumes?

### DISCUSSION THREAD

**Discussion by mattwarkentin** [TorchIO-project/torchio#740] [Q&A] [answered]
Hi @fepegar,

I am not sure if `torchio` has support for this, or if I need to do some pre-processing outside of `torchio` first. Basically, I have CT and binary mask file pairs. The masks were of course derived from the CT scan, but when the mask is saved to disk, for memory reasons, only the minimal bounding box containing the ROI is saved. The affine/origin is updated accordingly, so the sub-volume is "aware" it is contained within the CT, but from `torchio`'s perspective these are two volumes of different spatial shape and with different affines (at least the origin component).

Is there any good way to re-align these two volumes?

For additional context:
- CT: 512x512x322, spacing=[0.74, 0.74, 1], origin=[186.1, 333.1, -363.0]
- Mask: 35x35x26, spacing=[0.74, 0.74, 1], origin=[-49.8, 86.8, -111.0]

Hoping there is a good way to geometrically align based on their origins, and then pad the mask with 0's until it matches the spatial shape of the CT, and then updating them to have the same origin/affine.

**Comment by fepegar** [comment, marked as answer]
Hey @mattwarkentin.
1. Are you compressing those masks? The file size should be just some KB
2. You can use `Resample('ct')` to match the FOVs

**Reply by fepegar** [reply]
For example, I can simulate a pair of your images like this:

```python
import torchio as tio
subject = tio.datasets.BrainTumor()
subject.remove_image('t1')
subject.remove_image('t1c')
subject.remove_image('t2')
cropped = tio.CropOrPad(mask_name='seg')(subject)
subject.remove_image('seg')
subject.add_image(cropped.seg, 'seg')
subject.plot()
```

![Figure_1](https://user-images.githubusercontent.com/12688084/142011659-8855a4ca-86d0-44ef-89ce-8ba0592e09fb.png)

And then

```python
tio.Resample('flair')(subject).plot()
```

![Figure_2](https://user-images.githubusercontent.com/12688084/142011821-5afce0f2-f332-4792-8ba4-a8bf0f249b03.png)

**Reply by fepegar** [reply]
@OeslleLucena, this is what we did to your CT slices.

**Reply by mattwarkentin** [reply]
Torchio is so great. Thanks. Whenever I get frustrated with MONAI, I end up finding comfort with Torchio.

**Reply by mattwarkentin** [reply]
> Are you compressing those masks? The file size should be just some KB

No compression.

**Reply by fepegar** [reply]
No worries! Thanks for the kind words :D

**Reply by fepegar** [reply]
> No compression.

Any good reason for this? Just curious.

**Reply by mattwarkentin** [reply]
Good reasons for not compressing? Or good reasons for storing the masks as minimal bounding boxes?

**Reply by fepegar** [reply]
I guess it's the kind of the same question. My point is that if you compressed the masks, the file size would be quite small and you wouldn't need to care about the bounding boxes and use `Resample`. On the other hand, using smaller and uncompressed FOVs would optimize data loading, which I guess is better if you have a lot of storage but want things to be very fast.

**Reply by mattwarkentin** [reply]
It's actually just a consequence of the particular software used for doing the ROI segmentations. By default, it saves masks as minimal bounding boxes. I would've preferred it maintained the full geometry and just used compression instead (or nothing at all, the file sizes aren't that large compared to the CT).

**Reply by fepegar** [reply]
Thanks!

**Comment by romainVala** [comment]
Hi there, I think compression if of course very efficient (especially for binary mask with a lot of zeros), so you spare space on the disk but not on memory when you load those volumes.

I think this issue is related to #734 , it is exactly the same (sub volume) but the reason is that otherwise it does not fit into ram memory
